File: tests/state.test.ts
import { describe, expect, test } from 'bun:test';
import { mkdtempSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { loadState, recordUse, saveState } from '../src/state';

// state.json is a disposable per-Command cache, keyed by Command id —
// unreadable means silently reset to {}. lastUsedAt drives MRU; args are last
// Placeholder values.

const dir = () => mkdtempSync(join(tmpdir(), 'potato-state-'));

describe('loadState', () => {
  test('missing file is an empty state', () => {
    expect(loadState(join(dir(), 'state.json'))).toEqual({});
  });

  test('corrupt file silently resets to empty', () => {
    const file = join(dir(), 'state.json');
    writeFileSync(file, '{ nope');
    expect(loadState(file)).toEqual({});
  });

  test('non-object content silently resets to empty', () => {
    const file = join(dir(), 'state.json');
    writeFileSync(file, '[1, 2]');
    expect(loadState(file)).toEqual({});
  });
});

describe('recordUse + save/load round trip', () => {
  test('stores lastUsedAt and merges arg values per command id', () => {
    const file = join(dir(), 'state.json');
    let state = loadState(file);
    state = recordUse(state, 'cmd-1', { host: 'prod-2' }, new Date('2026-07-24T09:12:00Z'));
    saveState(file, state);
    const loaded = loadState(file);
    expect(loaded['cmd-1']!.lastUsedAt).toBe('2026-07-24T09:12:00.000Z');
    expect(loaded['cmd-1']!.args).toEqual({ host: 'prod-2' });
  });

  test('a later use updates the timestamp and merges new args over old', () => {
    let state = recordUse({}, 'x', { a: '1', b: '2' }, new Date('2026-01-01T00:00:00Z'));
    state = recordUse(state, 'x', { b: '3' }, new Date('2026-02-01T00:00:00Z'));
    expect(state['x']!.lastUsedAt).toBe('2026-02-01T00:00:00.000Z');
    expect(state['x']!.args).toEqual({ a: '1', b: '3' });
  });
});

File: internal/state/state_test.go
package state

import (
	"os"
	"path/filepath"
	"testing"
	"time"
)

// state.json is a disposable per-Command cache, keyed by Command id —
// unreadable means silently reset to empty. LastUsedAt drives MRU; Args are
// the last Placeholder values.

func TestLoadMissingFile(t *testing.T) {
	if s := Load(filepath.Join(t.TempDir(), "state.json")); len(s) != 0 {
		t.Errorf("got %v, want empty", s)
	}
}

func TestLoadCorruptFileResets(t *testing.T) {
	for _, content := range []string{"{ nope", "[1, 2]"} {
		file := filepath.Join(t.TempDir(), "state.json")
		if err := os.WriteFile(file, []byte(content), 0o644); err != nil {
			t.Fatal(err)
		}
		if s := Load(file); len(s) != 0 {
			t.Errorf("%q loaded as %v, want empty", content, s)
		}
	}
}

func TestRecordUseRoundTrip(t *testing.T) {
	file := filepath.Join(t.TempDir(), "state.json")
	s := Load(file)
	s = RecordUse(s, "cmd-1", map[string]string{"host": "prod-2"}, time.Date(2026, 7, 24, 9, 12, 0, 0, time.UTC))
	if err := Save(file, s); err != nil {
		t.Fatal(err)
	}
	loaded := Load(file)
	if loaded["cmd-1"].LastUsedAt != "2026-07-24T09:12:00.000Z" {
		t.Errorf("lastUsedAt = %q", loaded["cmd-1"].LastUsedAt)
	}
	if loaded["cmd-1"].Args["host"] != "prod-2" {
		t.Errorf("args = %v", loaded["cmd-1"].Args)
	}
}

func TestRecordUseMergesArgs(t *testing.T) {
	s := RecordUse(State{}, "x", map[string]string{"a": "1", "b": "2"}, time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC))
	s = RecordUse(s, "x", map[string]string{"b": "3"}, time.Date(2026, 2, 1, 0, 0, 0, 0, time.UTC))
	if s["x"].LastUsedAt != "2026-02-01T00:00:00.000Z" {
		t.Errorf("lastUsedAt = %q", s["x"].LastUsedAt)
	}
	if s["x"].Args["a"] != "1" || s["x"].Args["b"] != "3" {
		t.Errorf("args = %v, want a=1 b=3", s["x"].Args)
	}
}
